flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/AbstractStreamTaskNetworkInput.java

```diff
@@ -151,7 +152,11 @@ protected DataInputStatus processEvent(BufferOrEvent bufferOrEvent) {
         final AbstractEvent event = bufferOrEvent.getEvent();
         // TODO: with checkpointedInputGate.isFinished() we might not need to support any events on
         // this level.
-        if (event.getClass() == EndOfPartitionEvent.class) {
+        if (event.getClass() == EndOfData.class) {
+            if (checkpointedInputGate.hasReceivedEndOfData()) {
+                return DataInputStatus.END_OF_DATA;
+            }
+        } else if (event.getClass() == EndOfPartitionEvent.class) {
```

Comment:
move `releaseDeserializer()` to `EndOfData` handling?

Reply:
let's try it...

Reply:
I am afraid this would require more thought and careful investigation.

Moving it to `EndOfData` makes e.g. OrcFileSystemITCase or HiveSinkCompactionITCase fail occasionally. Unfortunately, I can not tell why is that, at least yet. Given this is just an optimization for an early resource release. Would you be ok with leaving it as it is for now and getting back to it at a later point?